auth0/provider.go

```diff
@@ -20,17 +27,28 @@ func init() {
 			"domain": {
 				Type:        schema.TypeString,
 				Required:    true,
-				DefaultFunc: schema.EnvDefaultFunc("AUTH0_DOMAIN", nil),
+				DefaultFunc: schema.EnvDefaultFunc(AUTH0_DOMAIN, nil),
 			},
 			"client_id": {
-				Type:        schema.TypeString,
-				Required:    true,
-				DefaultFunc: schema.EnvDefaultFunc("AUTH0_CLIENT_ID", nil),
+				Type:          schema.TypeString,
+				Optional:      true,
+				DefaultFunc:   schema.EnvDefaultFunc(AUTH0_CLIENT_ID, nil),
+				RequiredWith:  []string{"client_id", "client_secret"},
+				ConflictsWith: []string{"auth0_token"},
+				AtLeastOneOf:  []string{"client_id", "client_secret", "auth0_token"},
```

Comment:
I'm not sure if the `AtLeastOneOf` declarative rule is applicable to this provider because it's not exactly true. The provider needs at least a token or both ID and secret. 

I think the idea with these rules is if they apply to our provider, we should use them, but if not, we shouldn't try to forcibly adopt. So for that reason, I'd like to suggest we remove this one.